p4-16/spec/P4-16-spec.mdk

```diff
@@ -7097,16 +7099,157 @@ annotationBody
     | annotationBody annotationToken
 ~ End P4Grammar
 
-Annotations may impose additional structure on their bodies, and are not
-confined to the P4 language. For example, the P4Runtime specification defines a
-`@pkginfo` annotation that takes key-value pairs, which are not found in the
-main P4 language.
+Unstructured annotations may impose additional structure on their bodies, and
+are not confined to the P4 language. For example, the P4Runtime specification
+defines a `@pkginfo` annotation that takes key-value pairs, which are not found
+in the main P4 language.
+
+## Bodies of Structured Annotations
+
+Unlike unstructured annotations, structured annotations use square brackets
+`[...]` and have a specific format. They are commonly used to declare custom
+metadata, consisting of expression lists or key-value lists but not both (except
+that the value of a `kvPair` may itelf be an `expressionList`). An
+`expressionList` may be empty or contain a comma-separated list of members. A
+`kvList` consists of one or more `kvPair`s, each consisting of a key and a
+value; the value is an `expressionList`. 
+
+To avoid ambiguity, an empty `stucturedAnnotationBody` is by definition
+considered to contain an empty `expressionList`. Furthermore, the value of a `kvPair` may
+be the empty `expressionList`, in which case the pair is considered to
+contain a key without a value.
+
+The `expression`s within an `expressionList` must evaluate at compile-time to
+string literals, infinite-precision integers or booleans. Note that P4Runtime
+information (P4Info) may stipulate additional restrictions. For example, an
+integer expression might be limited to 64-bit values.
+
+It is illegal to duplicate a key within an annotation.  It is illegal to use the
+same annotation `name` on a given `annotationToken`.
+
+~ Begin P4Grammar
+stucturedAnnotationBody
+    : expressionList
+    | kvList
+    ;	
+...
+expressionList
+    : /* empty */
+    | expression
+    | expressionList ',' expression
+    ;
+...
+kvList
+    : kvPair
+    | kvList ',' kvPair
+    ;
+
+kvPair
+    : name '=' expressionList
+    ;
+
+~ End P4Grammar
+
+### Structured Annotation Examples
+
+#### Empty Expression List
```

Comment:
I don't think that a subsection is warranted for each example.

Reply:
Sure, I can remove the headings.